### PR TITLE
Add select_region method for event lists

### PR DESCRIPTION
**Pull request by registerrier**
Following PIG 10, see #2129 this PR introduces an `EventListBase.select_region(region, wcs)` as well as some tests.

This will be used for PR #2089 to properly select events.

### CODE REVIEW

**Comment by registerrier** [comment]
OK. I am implementing the change. 
One question though. At the moment `make_region(region)` can return anything if `region` is neither a `str` nor a `SkyRegion`. Wouldn't it be safer to raise an exception if `region` is neither a `str` nor a `SkyRegion`/`PixelRegion`?

**Comment by cdeil** [comment]
@registerrier - Yes, `make_region` should end like `make_pixel_region`, i.e. like this:
```
 elif isinstance(region, Region):
        return region
    else:
        raise TypeError("Invalid type: {!r}".format(region))
```

File: gammapy/data/event_list.py [hunk 1]
```diff
@@ -379,6 +380,45 @@ def filter_circular_region(self, region):
...
+            the world coordinate system transformation to assume
+
+        Returns
+        -------
+        index_array : `numpy.ndarray`
+            Index array of selected events
+        """
+        region = make_region(region)
+        position = self.radec
+        mask = np.where(region.contains(position, wcs))[0]
```
Comment (adonath) [inline gammapy/data/event_list.py]: I think there is no need for the call `np.where()`, just remove and remove a boolean array. Or is there an advantage in having the indices?

File: gammapy/data/event_list.py [hunk 2]
```diff
@@ -379,6 +380,45 @@ def filter_circular_region(self, region):
 
         return mask
 
+    def select_region(self, region, wcs):
+        """Select events in circular regions.
```
Comment (adonath) [inline gammapy/data/event_list.py]: I think `circular` is not true here, it's more general. Maybe change to `...in specified region.`?

File: gammapy/data/event_list.py [hunk 3]
```diff
@@ -379,6 +380,45 @@ def filter_circular_region(self, region):
...
+        Returns
+        -------
+        event_list : `EventList`
+            Copy of event list with selection applied.
+        """
+        mask = self.filter_region(region, wcs)
+        return self.select_row_subset(mask)
+
+    def filter_region(self, region, wcs):
+        """Create selection mask for event in given circular regions.
```
Comment (adonath) [inline gammapy/data/event_list.py]: "event in given circular regions" -> "events in given region."

File: gammapy/data/event_list.py [hunk 4]
```diff
@@ -379,6 +380,45 @@ def filter_circular_region(self, region):
...
+
+        Parameters
+        ----------
+        region : `~regions.SkyRegion` or str
+            Sky region or string defining a sky region
+        wcs : `~astropy.wcs.WCS`
+            the world coordinate system transformation to assume
```
Comment (adonath) [inline gammapy/data/event_list.py]: Start upper case...

File: gammapy/data/event_list.py (hunk 1 above)
Comment (registerrier) [reply gammapy/data/event_list.py]: Right.
In reality, there is probably no need to have `filter_region` and `select_region` at the same time.  This is copied from the `filter_circular_region` and `select_circular_region` logic. But this is probably irrelevant. `filter_circular_region` is never used in the code.

File: gammapy/data/event_list.py (hunk 2 above)
Comment (registerrier) [reply gammapy/data/event_list.py]: done

File: gammapy/data/event_list.py (hunk 3 above)
Comment (registerrier) [reply gammapy/data/event_list.py]: done

File: gammapy/data/event_list.py (hunk 4 above)
Comment (registerrier) [reply gammapy/data/event_list.py]: done

**Comment by registerrier** [comment]
I propose we adapt/remove the `filter_region` and `filter_circular_region` at the same time, once we decide what we do.
The travis CI errors are unrelated.
Merging now.

**Comment by cdeil** [comment]
@registerrier - Small follow-up commit in master in 880b10312 .

filter region is not a good name for a method that creates a mask. We should probably rename to `region_select_mask` so that it shows up in the API docs right next to the closely related `region_select` method. Or just remove those filter methods, because it's just two lines and thus not very useful, the few callers that need it can just put those two lines.
On the other hand, there are valid use cases where one needs the mask and then do other things with the mask than to apply it for row selection.

`filter_circular_region` is never called or tested and can just be deleted?

@registerrier @adonath - What's your preference concerning improving or removing the "filter" methods?

**Comment by adonath** [comment]
@cdeil I have a slight preference to just remove the `filter_...` methods. Numpy masking is really fundamental and I think it's OK for users to do this themselves.

**Pull request merged**